infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java

```diff
@@ -0,0 +1,70 @@
+package org.corfudb.infrastructure.logreplication.infrastructure;
+
+import org.corfudb.infrastructure.logreplication.infrastructure.plugins.ILogReplicationVersionAdapter;
+import org.corfudb.runtime.collections.TxnContext;
+
+/**
+ * Rolling upgrade handling means cluster must function in a mode where not all
+ * nodes are running the same codebase.
+ * Here the newer code base must function in a backward compatible mode until all
+ * the nodes have been upgraded to the same version.
+ * Once all nodes have upgraded it may atomically migrate data, clean up old code and switch
+ * completely to newer logic.
+ *
+ * To facilitate this in LR we add this module which is intended to be used as follows
+ * on LR startup and when upgrade is complete.
+ *
+ * For anything post corfu-0.4.0.1 we consider the code base to be at "V2"
+ * Anything prior to and including corfu-0.4.0.1 is V1
+ * Any code shipped that uses Source/Sink can be thought of as "V2"
+ *                  ┌───────────────────────────────────┐
+ *                  │ isClusterAllAtV2 = false (in-mem) │
+ *                  └──────────────────┬────────────────┘
+ *                                     │
+ *                     ┌───────────────▼───────────┐
+ *                   ┌─┤                           ├─┐
+ *         Yes ┌─────┴─┴┐  isRollingUpgradeOn() ? ┌┴─┴──────┐ No
+ *             │        └─────────────────────────┘         │
+ *             │                                            │
+ * ┌───────────▼──────────────┐                         ┌──────▼──────────────────┐
+ * │ isClusterAllAtV2 = false │                         │ isClusterAllAtV2 = true │
+ * └───────────┬──────────────┘                         └───────┬─────────────────┘
+ *             │            ┌──────────────────────┐            │
+ *             └───────────►│ start Log Replicator │◄───────────┘
+ *                          └──────────────────────┘
+ */
+public class LRRollingUpgradeHandler {
+
+    private volatile boolean isClusterAllAtV2 = false;
+    ILogReplicationVersionAdapter versionAdapter;
+    public LRRollingUpgradeHandler(ILogReplicationVersionAdapter versionAdapter) {
+        this.versionAdapter = versionAdapter;
+    }
+
+    public boolean isLRUpgradeInProgress(TxnContext txnContext) {
+        if (isClusterAllAtV2) {
+            return false;
+        }
+        String nodeVersion = versionAdapter.getNodeVersion();
+        String pinnedClusterVersion = versionAdapter.getPinnedClusterVersion(txnContext);
+        boolean isClusterUpgradeInProgress = !nodeVersion.equals(pinnedClusterVersion);
```

Comment:
This check may be sufficient in this version, but it will not work in subsequent versions.

Think about V3..
node 1 version = v2
node 2 version = v2
node 3 version = v3
cluster version = v2.

In this case, upgrade has not yet completed and we must not migrate the data

Reply:
but it is true for v2 right? it means the cluster is already at v2 same as what the current node is.
Also data migration will not be invoked in this case because when LR starts up , the v2 nodes will discover that the pinned version matches their own version and so will automatically bypass any checking right?